--- /home/ijntema/programming/securecopy/core-utils-based/V9.3.0-beta/coreutils-9.3/src/copy.h.org	2023-04-29 19:28:37.712015139 +0200
+++ /home/ijntema/programming/securecopy/core-utils-based/V9.3.0-beta/coreutils-9.3/src/copy.h	2023-04-29 19:28:37.716015014 +0200
@@ -110,20 +110,28 @@
    following programs: mv (when rename doesn't work), cp, install.
    So, if you add a new member, be sure to initialize it in
    mv.c, cp.c, and install.c.  */
 struct cp_options
 {
   enum backup_type backup_type;
 
   /* How to handle symlinks in the source.  */
   enum Dereference_symlink dereference;
 
+  /* HIJN */
+  /* Use CRC checksum to verify whether destination copy is same as source */
+  bool crc;
+
+  /* HIJN */
+  /* As crc, nut use xattr to get (fresh) crc or if empty/stale, update crc */
+  bool crc_with_xattr;
+
   /* This value is used to determine whether to prompt before removing
      each existing destination file.  It works differently depending on
      whether move_mode is set.  See code/comments in copy.c.  */
   enum Interactive interactive;
 
   /* Control creation of sparse files.  */
   enum Sparse_type sparse_mode;
 
   /* Set the mode of the destination file to exactly this value
      if SET_MODE is nonzero.  */
