--- /home/ijntema/programming/securecopy/core-utils-based/V9.3.0-beta/coreutils-9.3/m4/cu-progs.m4.org	2023-04-29 19:28:37.724014763 +0200
+++ /home/ijntema/programming/securecopy/core-utils-based/V9.3.0-beta/coreutils-9.3/m4/cu-progs.m4	2023-04-29 19:28:37.724014763 +0200
@@ -87,20 +87,23 @@
 gl_ADD_PROG([optional_bin_progs], [tsort])
 gl_ADD_PROG([optional_bin_progs], [tty])
 gl_ADD_PROG([optional_bin_progs], [uname])
 gl_ADD_PROG([optional_bin_progs], [unexpand])
 gl_ADD_PROG([optional_bin_progs], [uniq])
 gl_ADD_PROG([optional_bin_progs], [unlink])
 gl_ADD_PROG([optional_bin_progs], [vdir])
 gl_ADD_PROG([optional_bin_progs], [wc])
 gl_ADD_PROG([optional_bin_progs], [whoami])
 gl_ADD_PROG([optional_bin_progs], [yes])
+gl_ADD_PROG([optional_bin_progs], [crccp])
+gl_ADD_PROG([optional_bin_progs], [crcsum])
+gl_ADD_PROG([optional_bin_progs], [crcmv])
 no_install_progs_default='arch coreutils hostname'
 # Given the name of a variable containing a space-separated
 # list of install-by-default programs and the actual list of
 # do-not-install-by-default programs, modify the former variable
 # to reflect any "do-install" and "don't-install" requests.
 # That is, add any program specified via --enable-install-program,
 # and remove any program specified via --enable-no-install-program.
 # Note how the second argument below is a literal, with ","
 # separators.  That is required due to the way the macro works,
 # and since the corresponding ./configure option argument is
